File: java/src/main/java/com/leetcode/crackcodeinterview/ZeroFiveZeroThree.java
package com.leetcode.crackcodeinterview;

/**
 * @author: John
 * @date: 2020-03-25 21:58
 * @description: 0503
 **/
public class ZeroFiveZeroThree {

//    public int reverseBits(int num) {
//        int max = 0;
//        int cnt = 0, cntPre = 0;
//        while (num) {
//
//        }
//    }

    public int reverseBits(int num) {
        int maxLen = 0, preLen = 0, curLen = 0, bits = 32;
        while (bits-- > 0) {
            if ((num & 1) == 0) {
                curLen -= preLen;
                preLen = curLen + 1;
            }
            curLen++;
            maxLen = Math.max(maxLen, curLen);
            num >>= 1;
        }
        return maxLen;
    }
}

File: leetcodecpp/leetcode/crackcodeinterview/ZeroFiveZeroThree.cpp
//
// Created by John on 2020/3/25.
//

#include <bitstring.h>
#include "iostream"

using namespace std;

class Solution{
public:
    int reverseBits(int num) {
        int max = 0;
        int cnt = 0, cntPre = 0;
        while (num) {
//            cout << bit_set(num) << endl;
            cout << bitset<32>(num) << endl;
            if (1 & num) {
                cnt++;
            } else {
                if (cntPre + cnt + 1 > max) {
                    max = cntPre + cnt + 1;
                }
                cntPre = cnt;
                cnt = 0;
            }
            num >>= 1;
        }
        if (cntPre + cnt + 1 > max) {
            max = cntPre + cnt + 1;
        }
        return max;
    }
};

//int main() {
//    Solution *solution = new Solution();
//    solution->reverseBits(1775);
//}